Make Base API URL Dynamic in the UI

logyball opened the pull request:
## What does this PR change?
* Adds the ability for the BASE_URL variable to be overridden (overrode?) for the opencost UI.  Changes are spread across:
  -  **Dockerfile** - updated to run an ENTRYPOINT script rather than simply `nginx`
  - **docker-entrypoint.sh** - a small script that find/replaces the `BASE_URL` value within `allocation.js`
  - **src/services/allocation.js** - BASE_URL is updated to no longer be static or set at compile time, it is now set to a placeholder value which is determined at boot time

## Does this PR relate to any other PRs?
* Not as far as I know

## How will this PR impact users?
* It should not impact users.  It will impact developers who would like to run the opencost UI in an ingress-fronted environment.  Default behavior should not change.

## Does this PR address any GitHub or Zendesk issues?
* Closes #1677

## How was this PR tested?
* [see the linked comment](https://github.com/opencost/opencost/pull/1700#issuecomment-1440661383)

## Does this PR require changes to documentation?
* Yes - the README has been updated.  I could add more detail if necessary.

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next Opencost release? If not, why not?
* I am not entirely sure what to add here.  I don't know if the UI is included in the opencost release cycle. Help?

avrodrigues5 committed:
interfaces in cloudcost item and kubecost assetset and asset needed for matching billed vs monitored resources

Signed-off-by: Alan Rodrigues <alanr5691@yahoo.com>
Signed-off-by: Logan Ballard <loganballard@gmail.com>

kaelanspatel committed:
Fix incorrect value in node vgpu count assignment

Signed-off-by: Kaelan Patel <kaelanspatel@gmail.com>
Signed-off-by: Logan Ballard <loganballard@gmail.com>

avrodrigues5 committed:
return entire node instead of subset with lack of knowledge of return values

Signed-off-by: Alan Rodrigues <alanr5691@yahoo.com>
Signed-off-by: Logan Ballard <loganballard@gmail.com>

avrodrigues5 committed:
adding the get cluster functionality for a cloud cost item based on providers

Signed-off-by: Alan Rodrigues <alanr5691@yahoo.com>
Signed-off-by: Logan Ballard <loganballard@gmail.com>

logyball committed:
Adding parcel-cache to gitignore
Adding Docker-entrypoint script to UI docker container to allow for building the UI container with the ability to override the BASE_URL
Modifying the dockerfile for the UI to containe the entrypoint and to no longer build the container with a BASE_URL baked in
Adding a Makefile to the UI project to ease development
Removing the default from the AllocationService within the UI to be a placeholder
Dependencies updated themselves in package-lock

Signed-off-by: Logan Ballard <loganballard@gmail.com>

logyball committed:
adding package-lock to commit

Signed-off-by: Logan Ballard <loganballard@gmail.com>

AjayTripathy commented on ui/src/services/allocation.js:
```diff
@@ -1,7 +1,7 @@
 
 class AllocationService {
-  BASE_URL = process.env.BASE_URL || 'http://localhost:9090/model';
+  BASE_URL = process.env.BASE_URL || '{PLACEHOLDER_BASE_URL}';
```
Think that we want to make the default process.env.BASE_URL overrides PLACEHOLDER_BASE_URL overrides http://localhost:9090/model

logyball replied:
Hey @AjayTripathy I've updated the function call to now check for the PLACEHOLDER value and override with a default if it is not set.

I originally wanted to do something like:

`BASE_URL = process.env.BASE_URL || '{PLACEHOLDER_BASE_URL}' || 'http://localhost:9090/model';`

but realized that if the placeholder is not overwritten by the script, it will always be picked up as a valid value.  

I don't like that the conditional is checked _every_ API call, but it's definitely not a huge slowdown. 

My question to you (and any react experts): is there a way to run this check _once_ upon initializing the component, rather than doing it every time we want to make an API call?

AjayTripathy replied:
@nealormsbee any thoughts?

AjayTripathy commented:
Thanks so much for this PR! Know this is a draft but just want to make a quick comment + note that we want to make sure this is at least tested manually before we come out of Draft state!

thomasvn commented:
@logyball This looks good so far! Not sure if the `Makefile` is necessary here? I believe it's only used for a local build of the frontend?

logyball committed:
Updating script in dockerfile entrypoint to run parent containers entrypoint script
Adding nginx CMD to dockerfile
removing echos from makefile
adding local instructions to the README

Signed-off-by: Logan Ballard <loganballard@gmail.com>

logyball commented:
Sure @AjayTripathy , here's a couple of screenshots of the testing that I was doing.

Default behavior:

![image](https://user-images.githubusercontent.com/33067676/220723650-25134533-1c04-4217-b37c-03a4c9e402dd.png)

Overriding the default behavior locally:

![image](https://user-images.githubusercontent.com/33067676/220724473-579102d6-9992-4a02-a97e-33ae7d3c2983.png)

Docker image build from the Dockerfile without any overrides:

![image](https://user-images.githubusercontent.com/33067676/220733321-1baed3a6-54b6-4166-a53d-de5618379174.png)

Testing the docker image with an override:

![image](https://user-images.githubusercontent.com/33067676/220733603-dc63c00f-50ed-44e4-84a6-896d1f236745.png)

@thomasvn As far as the `makefile` goes,  I'd leave that call up to you guys! If you think it adds clutter, I can remove it, it's really just a convenience thing.  Looks like the UI is pretty minimal, and can probably have this kind of behavior replicated with just the `package.json`.  I can make changes there to prevent file bloat if you'd like.

logyball committed:
Merge branch 'develop' into logyball/make-ui-base-url-dynamic

logyball commented:
Sorry about the botched commit history, I keep forgetting to sign off my commits and have to rebase

Pokom commented:
@AjayTripathy I can't remember, do PR's generate images that we can run? If not, would you mind having an image generated for us? We can run this in our infra to test out as well.

AjayTripathy commented:
@mattray can you help @Pokom get an image of this to test against?

AjayTripathy commented:
Hey @logyball I did leave one comment about defaults on the code-- can you make that change and then I could go ahead and approve?

mattray commented:
@Pokom @logyball I rebased the PR to `develop` here: https://github.com/mattray/opencost/tree/logyball-make-ui-base-url-dynamic and built images for it here: https://hub.docker.com/layers/mattray/opencost-ui/logyball/images/sha256-a0fe3a5af4ecd7cc66513b154de9083dc6673baa2cbce72949740c7c8189e921?context=explore

I expect there's some way to build images for PRs automatically, but I bet it's pretty expensive for compute and hosting.

logyball committed:
Merge branch 'develop' into logyball/make-ui-base-url-dynamic

logyball committed:
updating readme to reflect how to override the base url

Signed-off-by: Logan Ballard <loganballard@gmail.com>

logyball commented:
Thanks @mattray for building that!

Hey guys I've got a couple remaining notes:
- merged in `develop` again, and pushed a new docker image for testing if you'd like https://hub.docker.com/layers/loganballard/opencost-ui/logan-test/images/sha256-7bbcb86a0020067117fffeead34c250d6191ce61d0a8f7f2e05f9acf3fb9e2ae?context=repo
  - can use `docker pull loganballard/opencost-ui:logan-test`
- Updated README to show how to override the base URL
- @AjayTripathy I'm not seeing any comments on the code, can you link me to them?
- @thomasvn would you like me to rework the `makefile` into the `package.json` to reduce clutter?

thomasvn commented:
@logyball We can keep the `Makefile` for now. Thanks for updating the `README.md` with details about it!

logyball commented:
Hey @AjayTripathy were you able to find those comments referenced above?

logyball committed:
Merge branch 'develop' into logyball/make-ui-base-url-dynamic

logyball committed:
Merge branch 'develop' into logyball/make-ui-base-url-dynamic

logyball committed:
Merge branch 'develop' into logyball/make-ui-base-url-dynamic

AjayTripathy commented:
Ahh super sorry @logyball I wrote a comment then forgot to hit 'submit review ' . Can you see it now?

logyball committed:
setting default API path if the placeholder is not overwritten

Signed-off-by: Logan Ballard <loganballard@gmail.com>

logyball committed:
porting makefile to package.json and updating readme

Signed-off-by: Logan Ballard <loganballard@gmail.com>

logyball committed:
updating readme and run scripts

Signed-off-by: Logan Ballard <loganballard@gmail.com>

logyball commented:
Ok responded to comments from @AjayTripathy  - please re-review when you have a chance.

@thomasvn  I know you said that the `makefile` was fine, but I was already in there and messing around and just decided to re-write it into the `package.json` so it's more familiar for `node` or `react` developers.  Let me know if you prefer the `makefile` (I personally do, but that's just me).  I also updated the README.

thomasvn commented:
@logyball I'm good with the `package.json` too! Thanks for updating the README accordingly!

logyball committed:
Merge branch 'develop' into logyball/make-ui-base-url-dynamic

logyball commented:
Here's the tests that I've run:

1. building and running docker image locally, without overriding `BASE_URL`: 
```sh
$ ❯ docker build -t loganballard/opencost-ui:logan-test .
[+] Building 81.5s (18/18) FINISHED                                                                              
 ... lots of other stuff...
$ ❯ docker run -p 9091:9090 -d loganballard/opencost-ui:logan-test
d989751e85de29a89f1ec94b2550aff0ae4748236780e54dd1624a7d722aaef0
$ ❯ curl localhost:9091/
<html gibberish>
```
![image](https://user-images.githubusercontent.com/33067676/223538383-d0c973a6-4c6e-443c-acf9-f9c7ab352a53.png)
2. Running docker image with BASE_URL override:
```sh
$ ❯ docker run --rm -e BASE_URL_OVERRIDE=/test/model -p 9091:9090 loganballard/opencost-ui:logan-test
d0e5f5781f15bfeba807235a881787f0496b5f03c0f890344077644a66d93f41
```
![image](https://user-images.githubusercontent.com/33067676/223541944-e094e975-aa29-4f18-9d66-e01d4703f3e5.png)
3. Running docker image with a FQDN as the base override:
```sh
$ ❯ docker run --rm -e BASE_URL_OVERRIDE=https://www.kubecost.com/ -p 9091:9090 loganballard/opencost-ui:logan-test
```
![image](https://user-images.githubusercontent.com/33067676/223542308-3b94c2d1-d71c-4724-9980-597956920d9c.png)
4. Building and running locally
```sh
$ ❯ npm run build                                         Node 18.3.0

> kubecost-ui-open@0.0.1 build
> npx parcel build src/index.html

✨ Built in 9.30s

dist/index.html                                   490 B    170ms
dist/favicon.7eff484d.ico                      24.54 KB     19ms
dist/index.89ba2f7f.css                        79.63 KB    266ms
dist/MaterialIcons-Regular.39de6e8b.eot        91.56 KB     61ms
dist/MaterialIcons-Regular.f6e8a8c0.woff2      80.56 KB    256ms
dist/MaterialIcons-Regular.7f7fc743.woff      103.25 KB     54ms
dist/MaterialIcons-Regular.0f326557.ttf       224.92 KB     18ms
dist/index.079c84fa.js                       1001.41 KB    4.71s
dist/logo.9a2faf1d.png                         10.68 KB     40ms
dist/index.b9855eff.js                        982.13 KB    4.80s
dist/logo.9a2faf1d.png                         10.68 KB     39ms

$ npm run serve
> kubecost-ui-open@0.0.1 serve
> npx parcel serve src/index.html

Server running at http://localhost:1234
✨ Built in 1.96s
```
5. Running locally with overridden URL
```sh
$ ❯ BASE_URL=http://localhost:9090/test npm run serve

> kubecost-ui-open@0.0.1 serve
> npx parcel serve src/index.html

Server running at http://localhost:1234
✨ Built in 772ms
```
![image](https://user-images.githubusercontent.com/33067676/223544078-e05f51c6-0d01-41a8-bd60-c495186ec437.png)

logyball committed:
removing equality check and replacing with contains check
equality check was overritten by overzealous entrypoint script

Signed-off-by: Logan Ballard <loganballard@gmail.com>

logyball commented:
Awesome thanks for approving @AjayTripathy - I don't have access to merge but if you're comfortable with it, feel free to send it!

mattray committed:
Merge branch 'develop' into logyball/make-ui-base-url-dynamic
